File: frontend/src/lib/payrollKe.ts
import { apiClient } from "@/lib/api";

export type PayFrequency = "MONTHLY" | "WEEKLY" | "BIWEEKLY";
export type PayrollRunStatus = "DRAFT" | "CALCULATED" | "APPROVED" | "PAID" | "CANCELLED";
export type KeComponentType = "NHIF" | "NSSF" | "HOUSING_LEVY" | "PERSONAL_RELIEF";
export type EmissionScope = "SCOPE1" | "SCOPE2" | "SCOPE3";

export interface PayrollProfile {
  id: string;
  employee_id: string;
  employee_name?: string;
  employee_code?: string;
  basic_salary: number;
  pay_frequency: PayFrequency;
  housing_allowance: number;
  transport_allowance: number;
  other_allowances?: Record<string, number>;
  tax_pin?: string;
  nhif_no?: string;
  nssf_no?: string;
  bank_name?: string;
  bank_account_no?: string;
  bank_branch?: string;
  is_active: boolean;
  notes?: string;
}

export interface TaxBand {
  id: string;
  tax_year: number;
  lower_limit: number;
  upper_limit?: number;
  rate: number;
  description?: string;
  is_active: boolean;
}

export interface StatutoryRate {
  id: string;
  component: KeComponentType;
  rate_year: number;
  rate_value: number;
  is_percentage: boolean;
  max_amount?: number;
  notes?: string;
  is_active: boolean;
}

export interface PayrollRun {
  id: string;
  run_no: string;
  period_month: number;
  period_year: number;
  period_start: string;
  period_end: string;
  run_date?: string;
  status: PayrollRunStatus;
  total_gross: number;
  total_paye: number;
  total_nhif: number;
  total_nssf: number;
  total_housing_levy: number;
  total_net: number;
  employee_count: number;
  notes?: string;
}

export interface PayrollLine {
  id: string;
  run_id: string;
  employee_id: string;
  employee_name?: string;
  employee_code?: string;
  tax_pin?: string;
  nhif_no?: string;
  nssf_no?: string;
  basic_salary: number;
  housing_allowance: number;
  transport_allowance: number;
  other_allowances: number;
  overtime_pay: number;
  gross_salary: number;
  nssf_employee: number;
  taxable_income: number;
  paye_before_relief: number;
  personal_relief: number;
  paye_amount: number;
  nhif_amount: number;
  housing_levy: number;
  other_deductions: number;
  total_deductions: number;
  net_salary: number;
  components_json?: {
    earnings: Record<string, number>;
    deductions: Record<string, number>;
  };
}

export interface Payslip {
  id: string;
  payroll_line_id: string;
  employee_id: string;
  run_id: string;
  issue_date: string;
  is_sent: boolean;
  employee_name?: string;
  employee_code?: string;
  department?: string;
  period_month?: number;
  period_year?: number;
  gross_salary?: number;
  net_salary?: number;
  paye_amount?: number;
  nhif_amount?: number;
  nssf_employee?: number;
  housing_levy?: number;
  components_json?: PayrollLine["components_json"];
}

export interface PayrollInsight {
  agent: string;
  insights: string[];
  details?: any[];
  generated_at: string;
}

export interface SeedResult {
  tax_bands_seeded: number;
  nhif_tiers_seeded: number;
  statutory_rates_seeded: number;
  message: string;
}

const BASE = "/api/v1/payroll-ke";

export const payrollKeApi = {
  // Profiles
  async listProfiles(): Promise<PayrollProfile[]> {
    const res = await apiClient.get(`${BASE}/profiles`);
    return res.data?.data ?? res.data;
  },
  async getProfile(id: string): Promise<PayrollProfile> {
    const res = await apiClient.get(`${BASE}/profiles/${id}`);
    return res.data?.data ?? res.data;
  },
  async createProfile(data: Omit<PayrollProfile, "id" | "employee_name" | "employee_code">): Promise<PayrollProfile> {
    const res = await apiClient.post(`${BASE}/profiles`, data);
    return res.data?.data ?? res.data;
  },
  async updateProfile(id: string, data: Partial<PayrollProfile>): Promise<PayrollProfile> {
    const res = await apiClient.patch(`${BASE}/profiles/${id}`, data);
    return res.data?.data ?? res.data;
  },

  // Tax bands & rates
  async listTaxBands(params?: { tax_year?: number }): Promise<TaxBand[]> {
    const res = await apiClient.get(`${BASE}/tax-bands`, { params });
    return res.data?.data ?? res.data;
  },
  async listStatutoryRates(): Promise<StatutoryRate[]> {
    const res = await apiClient.get(`${BASE}/statutory-rates`);
    return res.data?.data ?? res.data;
  },
  async seedRates(tax_year = 2024): Promise<SeedResult> {
    const res = await apiClient.post(`${BASE}/seed-rates`, null, { params: { tax_year } });
    return res.data?.data ?? res.data;
  },

  // Runs
  async listRuns(): Promise<PayrollRun[]> {
    const res = await apiClient.get(`${BASE}/runs`);
    return res.data?.data ?? res.data;
  },
  async getRun(id: string): Promise<PayrollRun> {
    const res = await apiClient.get(`${BASE}/runs/${id}`);
    return res.data?.data ?? res.data;
  },
  async createRun(data: { period_month: number; period_year: number; period_start: string; period_end: string; notes?: string }): Promise<PayrollRun> {
    const res = await apiClient.post(`${BASE}/runs`, data);
    return res.data?.data ?? res.data;
  },
  async calculateRun(id: string, tax_year?: number): Promise<PayrollRun> {
    const res = await apiClient.post(`${BASE}/runs/${id}/calculate`, null, { params: tax_year ? { tax_year } : {} });
    return res.data?.data ?? res.data;
  },
  async approveRun(id: string): Promise<PayrollRun> {
    const res = await apiClient.post(`${BASE}/runs/${id}/approve`);
    return res.data?.data ?? res.data;
  },
  async getRunLines(id: string): Promise<PayrollLine[]> {
    const res = await apiClient.get(`${BASE}/runs/${id}/lines`);
    return res.data?.data ?? res.data;
  },
  async getRunPayslips(id: string): Promise<Payslip[]> {
    const res = await apiClient.get(`${BASE}/runs/${id}/payslips`);
    return res.data?.data ?? res.data;
  },
  async getPayslip(id: string): Promise<Payslip> {
    const res = await apiClient.get(`${BASE}/payslips/${id}`);
    return res.data?.data ?? res.data;
  },

  // Reports
  async payeReport(runId: string) {
    const res = await apiClient.get(`${BASE}/runs/${runId}/reports/paye`);
    return res.data?.data ?? res.data;
  },
  async nhifReport(runId: string) {
    const res = await apiClient.get(`${BASE}/runs/${runId}/reports/nhif`);
    return res.data?.data ?? res.data;
  },
  async nssfReport(runId: string) {
    const res = await apiClient.get(`${BASE}/runs/${runId}/reports/nssf`);
    return res.data?.data ?? res.data;
  },
  async payrollSummary(runId: string) {
    const res = await apiClient.get(`${BASE}/runs/${runId}/reports/summary`);
    return res.data?.data ?? res.data;
  },

  // AI
  async aiAnomaly(runId: string): Promise<PayrollInsight> {
    const res = await apiClient.get(`${BASE}/runs/${runId}/ai/anomaly`);
    return res.data?.data ?? res.data;
  },
  async aiCompliance(): Promise<PayrollInsight> {
    const res = await apiClient.get(`${BASE}/ai/compliance`);
    return res.data?.data ?? res.data;
  },
};

export const MONTH_NAMES = [
  "", "January", "February", "March", "April", "May", "June",
  "July", "August", "September", "October", "November", "December",
];

export const STATUS_COLORS: Record<PayrollRunStatus, string> = {
  DRAFT: "bg-gray-100 text-gray-600",
  CALCULATED: "bg-blue-100 text-blue-700",
  APPROVED: "bg-green-100 text-green-700",
  PAID: "bg-purple-100 text-purple-700",
  CANCELLED: "bg-red-100 text-red-600",
};

File: backend/app/schemas/payroll_ke.py
from pydantic import BaseModel
from typing import Optional, List, Any
from decimal import Decimal
from datetime import datetime, date
import uuid

from app.models.payroll_ke import PayFrequency, KeComponentType, PayrollRunStatus


# ── Payroll Profile ───────────────────────────────────────────────────────────

class PayrollScopeFields(BaseModel):
    company_id: Optional[uuid.UUID] = None
    branch_id: Optional[uuid.UUID] = None
    department_id: Optional[str] = None
    cost_center_id: Optional[uuid.UUID] = None


class PayrollProfileCreate(PayrollScopeFields):
    employee_id: uuid.UUID
    basic_salary: Decimal
    pay_frequency: PayFrequency = PayFrequency.MONTHLY
    housing_allowance: Decimal = Decimal("0")
    transport_allowance: Decimal = Decimal("0")
    other_allowances: Optional[dict] = None
    tax_pin: Optional[str] = None
    nhif_no: Optional[str] = None
    nssf_no: Optional[str] = None
    bank_name: Optional[str] = None
    bank_account_no: Optional[str] = None
    bank_branch: Optional[str] = None
    is_active: bool = True
    notes: Optional[str] = None


class PayrollProfileUpdate(PayrollScopeFields):
    basic_salary: Optional[Decimal] = None
    housing_allowance: Optional[Decimal] = None
    transport_allowance: Optional[Decimal] = None
    other_allowances: Optional[dict] = None
    tax_pin: Optional[str] = None
    nhif_no: Optional[str] = None
    nssf_no: Optional[str] = None
    bank_name: Optional[str] = None
    bank_account_no: Optional[str] = None
    bank_branch: Optional[str] = None
    is_active: Optional[bool] = None
    notes: Optional[str] = None


class PayrollProfileRead(PayrollProfileCreate):
    id: uuid.UUID
    employee_name: Optional[str] = None
    employee_code: Optional[str] = None
    model_config = {"from_attributes": True}


# ── Tax Bands ─────────────────────────────────────────────────────────────────

class TaxBandCreate(BaseModel):
    tax_year: int
    lower_limit: Decimal
    upper_limit: Optional[Decimal] = None
    rate: Decimal
    effective_from: Optional[date] = None
    effective_to: Optional[date] = None
    description: Optional[str] = None
    is_active: bool = True


class TaxBandRead(TaxBandCreate):
    id: uuid.UUID
    model_config = {"from_attributes": True}


# ── Statutory Rates ───────────────────────────────────────────────────────────

class StatutoryRateCreate(BaseModel):
    component: KeComponentType
    rate_year: int
    rate_value: Decimal
    is_percentage: bool = True
    max_amount: Optional[Decimal] = None
    effective_from: Optional[date] = None
    effective_to: Optional[date] = None
    notes: Optional[str] = None
    is_active: bool = True


class StatutoryRateRead(StatutoryRateCreate):
    id: uuid.UUID
    model_config = {"from_attributes": True}


# ── Payroll Run ───────────────────────────────────────────────────────────────

class PayrollRunCreate(PayrollScopeFields):
    period_month: int
    period_year: int
    period_start: date
    period_end: date
    notes: Optional[str] = None


class PayrollRunRead(BaseModel):
    id: uuid.UUID
    run_no: str
    period_month: int
    period_year: int
    company_id: Optional[uuid.UUID] = None
    branch_id: Optional[uuid.UUID] = None
    department_id: Optional[str] = None
    cost_center_id: Optional[uuid.UUID] = None
    period_start: date
    period_end: date
    run_date: Optional[date] = None
    status: PayrollRunStatus
    total_gross: Decimal
    total_paye: Decimal
    total_nhif: Decimal
    total_nssf: Decimal
    total_housing_levy: Decimal
    total_net: Decimal
    employee_count: int
    notes: Optional[str] = None
    locked_at: Optional[datetime] = None
    locked_by_id: Optional[uuid.UUID] = None
    model_config = {"from_attributes": True}


# ── Payroll Lines ─────────────────────────────────────────────────────────────

class KePayrollLineRead(BaseModel):
    id: uuid.UUID
    run_id: uuid.UUID
    employee_id: uuid.UUID
    company_id: Optional[uuid.UUID] = None
    branch_id: Optional[uuid.UUID] = None
    department_id: Optional[str] = None
    cost_center_id: Optional[uuid.UUID] = None
    employee_name: Optional[str] = None
    employee_code: Optional[str] = None
    tax_pin: Optional[str] = None
    nhif_no: Optional[str] = None
    nssf_no: Optional[str] = None
    basic_salary: Decimal
    housing_allowance: Decimal
    transport_allowance: Decimal
    other_allowances: Decimal
    overtime_pay: Decimal
    gross_salary: Decimal
    nssf_employee: Decimal
    taxable_income: Decimal
    paye_before_relief: Decimal
    personal_relief: Decimal
    paye_amount: Decimal
    nhif_amount: Decimal
    housing_levy: Decimal
    other_deductions: Decimal
    total_deductions: Decimal
    net_salary: Decimal
    notes: Optional[str] = None
    components_json: Optional[dict] = None
    model_config = {"from_attributes": True}


# ── Payslip ───────────────────────────────────────────────────────────────────

class PayslipRead(BaseModel):
    id: uuid.UUID
    payroll_line_id: uuid.UUID
    employee_id: uuid.UUID
    company_id: Optional[uuid.UUID] = None
    branch_id: Optional[uuid.UUID] = None
    department_id: Optional[str] = None
    run_id: uuid.UUID
    issue_date: date
    is_sent: bool
    viewed_at: Optional[datetime] = None
    sent_at: Optional[datetime] = None
    sent_by_id: Optional[uuid.UUID] = None
    notes: Optional[str] = None
    employee_name: Optional[str] = None
    employee_code: Optional[str] = None
    department: Optional[str] = None
    period_month: Optional[int] = None
    period_year: Optional[int] = None
    gross_salary: Optional[Decimal] = None
    net_salary: Optional[Decimal] = None
    paye_amount: Optional[Decimal] = None
    nhif_amount: Optional[Decimal] = None
    nssf_employee: Optional[Decimal] = None
    housing_levy: Optional[Decimal] = None
    components_json: Optional[dict] = None
    model_config = {"from_attributes": True}


# ── Reports ───────────────────────────────────────────────────────────────────

class PayeSummaryRow(BaseModel):
    employee_name: str
    employee_code: str
    tax_pin: Optional[str] = None
    gross_salary: Decimal
    taxable_income: Decimal
    paye_amount: Decimal
    personal_relief: Decimal


class NhifSummaryRow(BaseModel):
    employee_name: str
    employee_code: str
    nhif_no: Optional[str] = None
    gross_salary: Decimal
    nhif_amount: Decimal


class NssfSummaryRow(BaseModel):
    employee_name: str
    employee_code: str
    nssf_no: Optional[str] = None
    gross_salary: Decimal
    nssf_employee: Decimal


class PayrollSummaryRow(BaseModel):
    employee_name: str
    employee_code: str
    department: Optional[str] = None
    gross_salary: Decimal
    total_deductions: Decimal
    net_salary: Decimal
    payment_method: Optional[str] = None


# ── AI ────────────────────────────────────────────────────────────────────────

class PayrollInsight(BaseModel):
    agent: str
    insights: List[str]
    details: Optional[List[dict]] = None
    generated_at: str


# ── Seed result ───────────────────────────────────────────────────────────────

class SeedResult(BaseModel):
    tax_bands_seeded: int
    nhif_tiers_seeded: int
    statutory_rates_seeded: int
    message: str
